src/promptflow-core/promptflow/executor/_async_nodes_scheduler.py

```diff
@@ -88,12 +89,16 @@ async def _execute_with_thread_pool(
         nodes: List[Node],
         inputs: Dict[str, Any],
         context: FlowExecutionContext,
+        timeout_seconds: Optional[int] = None,
     ) -> Tuple[dict, dict]:
         flow_logger.info(f"Start to run {len(nodes)} nodes with the current event loop.")
         dag_manager = DAGManager(nodes, inputs)
+        timeout_task = asyncio.create_task(asyncio.sleep(timeout_seconds)) if timeout_seconds else None
         task2nodes = self._execute_nodes(dag_manager, context, executor)
         while not dag_manager.completed():
-            task2nodes = await self._wait_and_complete_nodes(task2nodes, dag_manager)
+            task2nodes = await self._wait_and_complete_nodes(
+                task2nodes, dag_manager, context, timeout_task, timeout_seconds
+            )
```

Comment:
How about adding another method called `_cancel_nodes()` and then we modify the logic here to something like:

```python
while not dag_manager.completed():
    if execution_time >= timeout_seconds:
        await self._cancel_nodes()
        break

    # Do the exiting thing here, like:
    task2nodes = await self._wait_and_complete_nodes(task2nodes, dag_manager)
```

Reply:
updated